---
title: "MethylGenotyper"
author: "Minghan Qu, Yi Jiang, Chaolong Wang"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{MethylGenotyper}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

The `MethylGenotyper` package provides functions to infer genotypes (produce a VCF file) for specific probes and on Illumina's Infinium MethylationEPIC (EPIC) array. The probes include SNP probes, Type I CCS (color-channel-switching) probes, and Type II CCS probes. We defined RAI as the Ratio of Alternative allele Intensity to total intensity and calculated RAI for each probe and sample. 

- __SNP probe:__ There are 59 SNP probes (started with "rs") on EPIC array. Six of them on sex chromosomes were removed. We aligned each probe sequence to reference genome and calculated RAI, which is defined as the proportion of probe signals supporting alternative allele.
- __Type I CCS probe:__ Training from our in-house data (Chinese population), we provide a list of 132 Type I CCS probes on EPIC array. It's opt to you to train your own models. These probes are type I probes with SNPs at the extension bases resulting in CCS (A,T <-> C,G mutation). The signals for probes with CCS SNPs are called out-of-band signals. The RAI is defined as the proportion of out-of-band signals over total signals.
- __Type II CCS probe:__ For this probe type, we only focus on probes with SNPs at the extension bases (CpG target sites). Training from our in-house data (Chinese population), we provide a list of 4853 Type II CCS probes. It's opt to you to train your own models. The RAI is defined as the proportion of signals supporting alternative alleles over total signals.

The genotypes are called in two steps: 1) We fit a mixture of three beta distribution for the three genotypes (reference homozygous, heterozygous, and alternative homozygous) based on the Expectationâ€“maximization (EM) algorithm; 2) We use the Bayesian approach to get the genotype probabilities ($P_0$, $P_1$, and $P_2$ for reference homozygous, heterozygous, and alternative homozygous, respectively), with AFs of the matched population (the 1000 Genomes Project, 1KGP) being used as priors. A VCF file containing allele frequency (AF), $R^2$, genotype probability, hard genotype, and dosage genotype (DS) will be produced:

- __DS:__ $G = P_1 + 2P_2$
- __AF:__ $p = \frac{\sum G}{2N}$, where $N$ is the sample size
- __$R^2$:__ $R^2 = \frac{\sigma^2}{2p(1-p)}$, where $\sigma^2$ is the variance of dosage genotypes

As an option, we also provided several convenient tools for estimating kinships and sample contamination.

## Dependencies

This document has the following dependencies

```{r dependencies, eval=TRUE, message=FALSE, warning=FALSE}
library(MethylGenotyper)
library(minfi)
library(tidyverse)
library(foreach)
library(doParallel)
library(HardyWeinberg)
library(multimode)
library(rlist)
library(stats4)
```

## Recommended workflow

### Read manifest file and IDAT file list

Load EPIC manifest file. Required columns: Name, AddressA_ID, AddressB_ID, Infinium_Design_Type, and Color_Channel.

```{r mnfst, eval=TRUE, warning=FALSE, message=TRUE}
data(mnfst)
head(mnfst)
```

Read IDAT file list. Here is an example of processing three IDAT files from `minfiDataEPIC`. Note that this is just an exemplification of how this tool works. We strongly recommend to use a larger sample size to test the code, such as [GSE112179](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112179). Your may process your own data by specifying your target file list. Required collumns: Sample_Name, Basename.

```{r rgSet, eval=TRUE, warning=FALSE, message=TRUE}
target <- get_target("EPIC")
head(target)
```

### Read IDAT files and noob and dye-bias correction

With the following code, the IDAT files listed in the variable `target` will be read one-by-one. For each sample, a noob background correction and dye-bias correction will be done. You can specify the number of CPUs to enable parallel processing. After that, a list of four elements will be returned, including corrected signals of probe A and probe B for the two color channels.

```{r correct_noob_dye, eval=FALSE, warning=FALSE, message=FALSE}
rgData <- correct_noob_dye(target, mnfst, cpu=3)
```

### Call genotypes

As shown below, you can call genotypes for SNP probes, Type I CCS probes, and Type II CCS probes. For each probe and each sample, an RAI will be calculated, which indicates the proportion of signals supporting the alternative allele. The RAI values can be classified into three clusters by fitting a mixture of three beta distributions. Genotype probabilities can be calculated by using a Bayesian approach. In addition, you can plot the distribution of the RAI values and produce a VCF file of the inferred genotypes by specifying `plotBeta=TRUE` and `vcf=TRUE`.

You can also specify $R^2$ and MAF cutoffs to filter variants. Note that for VCF output, variants with $R^2$ or MAF below the cutoffs will be marked in the `FILTER` column. For the returned matrix, variants with $R^2$ or MAF below the cutoffs will be removed.

We noted that in the example data, most of variants have $R^2$=0. This is because we only used three samples here. We strongly recommend to use a larger sample size to test the code, such as [GSE112179](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112179).

```{r callGeno, eval=FALSE, warning=FALSE, message=FALSE}
# Call genotypes for SNP probes
genotype_snp <- callGeno_snp(rgData, input="raw", vcf=TRUE, pop="EAS")
# Call genotypes for Type I CCS probes
genotype_typeI <- callGeno_typeI(rgData, vcf=TRUE, pop="EAS")
# Call genotypes for Type II CCS probes
genotype_typeII <- callGeno_typeII(rgData, input="raw", vcf=TRUE, pop="EAS")
# Combine genotypes inferred from the three probe types
dosage <- rbind(genotype_snp$dosage, genotype_typeI$dosage, genotype_typeII$dosage)
```

As an alternative option, you can input a matrix of beta values or M values, with each row indicates a probe and each column indicates a sample. This option only works for SNP probes and Type II CCS probes. Here are the examples of calling genotypes from beta values. For input of M values, please specify `input="mval"`. Remember to conduct background correction and dye-bias correction before running the following code.

```{r callGeno2, eval=FALSE, warning=FALSE, message=FALSE}
# Call genotypes for SNP probes
genotype_snp <- callGeno_snp(beta_matrix, input="beta", vcf=TRUE, pop="EAS")
# Call genotypes for Type II CCS probes
genotype_typeII <- callGeno_typeII(beta_matrix, input="beta", vcf=TRUE, pop="EAS")
# Combine genotypes inferred from the three probe types
dosage <- rbind(genotype_snp$dosage, genotype_typeII$dosage)
```

### Estimate sample relationships

With the inferred genotypes, you can estimate sample relationships using the `getKinship` function. It produces a data frame containing kinship coefficient ($\phi$) and sample relationships between each two samples.

Kinship coefficient is calculated according to the SEEKIN software (Dou et al. Plos Genetics 2017):
$$2\phi_{ij} = \frac{\sum_m(G_{im}-2p_m)(G_{jm}-2p_m)}{2p_m(1-p_m)(R^2)^2}$$,
where $\phi_{ij}$ denotes the kinship coefficient between $i$-th and $j$-th sample. $G_{im}$ and $G_{jm}$ denotes genotypes of $m$-th SNP for $i$-th and $j$-th sample. $p_m$ denotes allele frequency of $m$-th SNP. $R^2$ is calculated as $R^2 = \frac{\sigma^2}{2p(1-p)}$, where $\sigma^2$ is the variance of genotype.

Sample relationships can be inferred according to kinship coefficients (Manichaikul et al. Bioinformatics 2010). Sample pairs with $\phi \ge \frac{1}{2^\frac{3}{2}}$ are duplicate samples or monozygotic twin. Sample pairs with $\phi < \frac{1}{2^\frac{3}{2}}$ and $\phi \ge \frac{1}{2^\frac{5}{2}}$ are first-degree relatives. Sample pairs with $\phi < \frac{1}{2^\frac{5}{2}}$ and $\phi \ge \frac{1}{2^\frac{7}{2}}$ are second-degree relatives. Sample pairs with $\phi < \frac{1}{2^\frac{7}{2}}$ and $\phi \ge \frac{1}{2^\frac{9}{2}}$ are third-degree relatives. Sample pairs with $\phi < \frac{1}{2^\frac{9}{2}}$ are unrelated. 

```{r getKinship, eval=FALSE, warning=FALSE, message=FALSE}
kinship <- getKinship(dosage)
```

### Estimate sample contamination

We recommend to set samples with inbreeding coefficients beyond 3 SDs of the mean as contaminated.

```{r getInbreed, eval=FALSE, warning=FALSE, message=FALSE}
inbreed <- getInbreed(dosage)
```

