<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Yi Jiang, Minghan Qu, Chaolong Wang" />

<meta name="date" content="2023-06-14" />

<title>MethylGenotyper</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>


<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">MethylGenotyper</h1>
<h4 class="author">Yi Jiang, Minghan Qu, Chaolong Wang</h4>
<h4 class="date">2023-06-14</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>The <code>MethylGenotyper</code> package provides functions to infer
genotypes (produce a VCF file) for specific probes and on Illumina
methylation array (EPIC or 450K). The probes include SNP probes, Type I
CCS (color-channel-switching) probes, and Type II CCS probes. We defined
RAI as the Ratio of Alternative allele Intensity to total intensity and
calculated RAI for each probe and sample.</p>
<ul>
<li><strong>SNP probe:</strong> There are 59 SNP probes (started with
“rs”) on EPIC array and 65 SNP probes on 450K array. Probes on sex
chromosomes (six in EPIC and eight on 450K) were removed. We aligned
each probe sequence to reference genome and calculated RAI, which is
defined as the proportion of probe signals supporting alternative
allele.</li>
<li><strong>Type I CCS probe:</strong> These probes are type I probes
with SNPs at the extension bases resulting in CCS (A,T &lt;-&gt; C,G
mutation). The signals for probes with CCS SNPs are called out-of-band
signals. The RAI is defined as the proportion of out-of-band signals
over total signals.</li>
<li><strong>Type II CCS probe:</strong> For this probe type, we only
focus on type II probes with SNPs at the extension bases (CpG target
sites), the reference alleles for the SNPs are C/G, and the alternative
alleles are A/T. The RAI is defined as the proportion of signals
supporting alternative alleles over total signals.</li>
</ul>
<p>The RAI values are usually distributed with three peaks, representing
the three genotypes (reference homozygous, heterozygous, and alternative
homozygous). To call genotypes from the RAI values, we fit a mixture of
three beta distributions for the three genotypes and a uniform
distribution for outliers based on the Expectation–maximization (EM)
algorithm. Probe-specific weights derived from allele frequencies (AFs)
were used in the EM algorithm. A VCF file containing dosage genotype
(<span class="math inline">\(\hat{D}_{ij}\)</span>), AF (<span class="math inline">\(\hat{q}_i\)</span>), and <span class="math inline">\(\hat{R}_i^2\)</span> will be produced:</p>
<ul>
<li><span class="math inline">\(\hat{D}_{ij} =
\sum_{k=0}^{2}{k\hat{P}_{ijk}}\)</span>, where <span class="math inline">\(\hat{P}_{ijk}\)</span> is the genotype
probability</li>
<li><span class="math inline">\(\hat{q}_i =
\frac{\sum_{j=1}^n\hat{D}_{ij}}{2n}\)</span>, where <span class="math inline">\(n\)</span> is the sample size</li>
<li><span class="math inline">\(\hat{R}_i^2 =
\frac{Var(\hat{D}_{ij})}{2\hat{q}_i(1-\hat{q}_i)}\)</span></li>
</ul>
<p>For samples of mixed population, we provided an option to infer
population structure and calculate individual-specific AFs, which can
improve the accuracy of estimating kinship coefficients.</p>
<p>We also provided functions to estimate kinship coefficients and
sample contamination.</p>
</div>
<div id="recommended-workflow" class="section level2">
<h2>Recommended workflow</h2>
<div id="load-the-methylgenotyper-package" class="section level3">
<h3>Load the MethylGenotyper package</h3>
<p>This package has the following dependencies: <code>minfi</code>,
<code>tidyverse</code>, <code>foreach</code>, <code>doParallel</code>,
<code>HardyWeinberg</code>, <code>multimode</code>, <code>rlist</code>,
<code>stats4</code>, <code>ggplot2</code>, <code>ggpubr</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MethylGenotyper)</span></code></pre></div>
</div>
<div id="read-idat-files-and-perform-noob-and-dye-bias-correction" class="section level3">
<h3>Read IDAT files and perform noob and dye-bias correction</h3>
<p>Read IDAT file list. Here is an example of processing three IDAT
files from <code>minfiDataEPIC</code>. Note that this is just an
exemplification of how this tool works. We strongly recommend to use a
larger sample size to test the code, such as <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112179">GSE112179</a>.
Your may process your own data by specifying your target file list.
Required collumns: Sample_Name, Basename.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>target <span class="ot">&lt;-</span> <span class="fu">get_target</span>(<span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(target)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    Sample_Name</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 200144450018</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 200144450019</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 200144450021</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                                                                           Basename</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 C:/Users/DELL/Documents/R/win-library/4.0/minfiDataEPIC/extdata/200144450018/200144450018_R04C01</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 C:/Users/DELL/Documents/R/win-library/4.0/minfiDataEPIC/extdata/200144450019/200144450019_R07C01</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 C:/Users/DELL/Documents/R/win-library/4.0/minfiDataEPIC/extdata/200144450021/200144450021_R05C01</span></span></code></pre></div>
<p>With the following code, the IDAT files listed in <code>target</code>
will be read one-by-one. For each sample, a noob background correction
and dye-bias correction will be conducted. You can specify the number of
CPUs to enable parallel processing. After that, a list of four elements
will be returned, including corrected signals of probe A and probe B for
the two color channels.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>rgData <span class="ot">&lt;-</span> <span class="fu">correct_noob_dye</span>(target, <span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>, <span class="at">cpu=</span><span class="dv">3</span>)</span></code></pre></div>
</div>
<div id="call-genotypes" class="section level3">
<h3>Call genotypes</h3>
<p>The genotype-calling procedure can be done for SNP probes, Type I CCS
probes, and Type II CCS probes, separately. You should specify the
correct population (One of EAS, AMR, AFR, EUR, SAS, and ALL) for your
samples to get accurate genotype calls. If you have samples of mixed
population, please specify the population with the largest sample
size.</p>
<p>You can plot the distribution of the RAI values and produce a VCF
file of the inferred genotypes by specifying <code>plotBeta=TRUE</code>
and <code>vcf=TRUE</code>.</p>
<p>You can also specify cutoffs of <span class="math inline">\(R^2\)</span>, MAF, HWE, and missing rate to filter
variants. Note that for VCF output, variants beyond the cutoffs will be
marked in the <code>FILTER</code> column.</p>
<p>We noted that in the example data, most of variants have <span class="math inline">\(R^2\)</span>=0. This is because we only used three
samples here. We strongly recommend to use a larger sample size to test
the code, such as <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE112179">GSE112179</a>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call genotypes for SNP probes, Type I CCS probes, and Type II CCS probes</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>genotype_snp <span class="ot">&lt;-</span> <span class="fu">callGeno_snp</span>(rgData, <span class="at">input=</span><span class="st">&quot;raw&quot;</span>, <span class="at">vcf=</span><span class="cn">TRUE</span>, <span class="at">pop=</span><span class="st">&quot;EAS&quot;</span>, <span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>genotype_typeI <span class="ot">&lt;-</span> <span class="fu">callGeno_typeI</span>(rgData, <span class="at">vcf=</span><span class="cn">TRUE</span>, <span class="at">pop=</span><span class="st">&quot;EAS&quot;</span>, <span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>genotype_typeII <span class="ot">&lt;-</span> <span class="fu">callGeno_typeII</span>(rgData, <span class="at">input=</span><span class="st">&quot;raw&quot;</span>, <span class="at">vcf=</span><span class="cn">TRUE</span>, <span class="at">pop=</span><span class="st">&quot;EAS&quot;</span>, <span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine genotypes inferred from the three probe types</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>dosage <span class="ot">&lt;-</span> <span class="fu">rbind</span>(genotype_snp<span class="sc">$</span>dosage, genotype_typeI<span class="sc">$</span>dosage, genotype_typeII<span class="sc">$</span>dosage)</span></code></pre></div>
<p>As an alternative option, you can input a matrix of beta values or M
values, with each row indicates a probe and each column indicates a
sample. This option only works for SNP probes and Type II CCS probes.
Here are the examples of calling genotypes from beta values. For input
of M values, please specify <code>input=&quot;mval&quot;</code>. Remember to
conduct background correction and dye-bias correction before running the
following code. Also be noted that other correction should NOT be
conducted, like BMIQ, as it flattens the peaks through a scale
transformation.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Call genotypes for SNP probes and Type II CCS probes</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>genotype_snp <span class="ot">&lt;-</span> <span class="fu">callGeno_snp</span>(beta_matrix, <span class="at">input=</span><span class="st">&quot;beta&quot;</span>, <span class="at">vcf=</span><span class="cn">TRUE</span>, <span class="at">pop=</span><span class="st">&quot;EAS&quot;</span>, <span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>genotype_typeII <span class="ot">&lt;-</span> <span class="fu">callGeno_typeII</span>(beta_matrix, <span class="at">input=</span><span class="st">&quot;beta&quot;</span>, <span class="at">vcf=</span><span class="cn">TRUE</span>, <span class="at">pop=</span><span class="st">&quot;EAS&quot;</span>, <span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine genotypes inferred from the three probe types</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>dosage <span class="ot">&lt;-</span> <span class="fu">rbind</span>(genotype_snp<span class="sc">$</span>dosage, genotype_typeII<span class="sc">$</span>dosage)</span></code></pre></div>
</div>
<div id="infer-population-structure-and-individual-specific-afs-for-mixed-population" class="section level3">
<h3>Infer population structure and individual-specific AFs for mixed
population</h3>
<p><strong>Project the study samples to reference ancestral
space:</strong> Principal Components Analyses (PCA) are conducted in
1KGP individuals (the ancestral space) and a combination of 1KGP
individuals and each study sample. Projection Procrustes analyses are
then conducted to project each study sample to reference ancestral
space. This step was originally implemented by the TRACE software and we
have adapted it in R (<a href="https://www.nature.com/articles/ng.2924">Wang et al. Nat Genet
2014</a>, <a href="http://dx.doi.org/10.1016/j.ajhg.2015.04.018">Wang et
al. Am J Hum Genet 2015</a>).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA and Procrustes analysis, based on genotypes of all probes passing QC</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>pc <span class="ot">&lt;-</span> <span class="fu">projection</span>(dosage, <span class="at">plotPCA=</span><span class="cn">TRUE</span>, <span class="at">cpu=</span><span class="dv">3</span>, <span class="at">platform=</span><span class="st">&quot;EPIC&quot;</span>)</span></code></pre></div>
<p><strong>Estimate individual-specific AFs:</strong> For each SNP, we
model genotypes of the reference individuals as a linear function of top
four PCs (<span class="math inline">\(v_\cdot\)</span>): <span class="math inline">\(G\sim\beta_0+\beta_1v_1+\beta_2v_2+\beta_3v_3+\beta_4v_4\)</span>.
Then, the individual AF (<span class="math inline">\(q\)</span>) for
each SNP and each sample can be obtained by: <span class="math inline">\(\hat{q}=\frac{1}{2}(\hat{\beta_0}+\hat{\beta_1}\hat{v_1}+\hat{\beta_2}\hat{v_2}+\hat{\beta_3}\hat{v_3}+\hat{\beta_4}\hat{v_4})\)</span>,
where <span class="math inline">\(\hat{v}_\cdot\)</span> are top four
PCs in study samples. (<a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007021">Dou
et al. PLoS Genet 2017</a>)</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(cpg2snp)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>snpvec <span class="ot">&lt;-</span> cpg2snp[<span class="fu">c</span>(<span class="fu">rownames</span>(genotype_snp<span class="sc">$</span>genotypes<span class="sc">$</span>RAI), <span class="fu">rownames</span>(genotype_typeI<span class="sc">$</span>genotypes<span class="sc">$</span>RAI), <span class="fu">rownames</span>(genotype_typeII<span class="sc">$</span>genotypes<span class="sc">$</span>RAI))]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>indAF <span class="ot">&lt;-</span> <span class="fu">get_indAF</span>(snpvec, pc<span class="sc">$</span>refPC, pc<span class="sc">$</span>studyPC)</span></code></pre></div>
<!-- **Recalibrate genotype probabilities based on individual-specific AFs: ** This step is based on the Bayesian approach, which can be implemented by using the `get_GP_bayesian` function. -->
<!-- ```{r recal_geno, eval=FALSE, warning=FALSE, message=FALSE} -->
<!-- # Recalibrate genotypes for SNP probes, Type I CCS probes, and Type II CCS probes -->
<!-- genotype_snp_recal <- recal_Geno(genotype_snp, type="snp_probe", indAF=indAF, platform="EPIC") -->
<!-- genotype_typeI_recal <- recal_Geno(genotype_typeI, type="typeI_ccs_probe", indAF=indAF, platform="EPIC") -->
<!-- genotype_typeII_recal <- recal_Geno(genotype_typeII, type="typeII_ccs_probe", indAF=indAF, platform="EPIC") -->
<!-- # Combine genotypes inferred from the three probe types -->
<!-- dosage_recal <- rbind(genotype_snp_recal$dosage, genotype_typeI_recal$dosage, genotype_typeII_recal$dosage) -->
<!-- ``` -->
</div>
<div id="estimate-sample-relationships-and-sample-contamination" class="section level3">
<h3>Estimate sample relationships and sample contamination</h3>
<p>With the inferred genotypes, you can estimate sample relationships
and sample contamination using the <code>getKinship</code> function. The
output of this function is a list of two elements: 1) a data frame
containing kinship coefficient (<span class="math inline">\(\phi\)</span>) and sample relationships between
each two samples; 2) a vector of inbreeding coefficients, which can be
used to infer sample contamination.</p>
<p>Kinship coefficient is calculated according to the SEEKIN software
(<a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1007021">Dou
et al. PLoS Genet 2017</a>): <span class="math display">\[2\phi_{ij} =
\frac{\sum_m(G_{im}-2p_m)(G_{jm}-2p_m)}{2p_m(1-p_m)(R^2)^2}\]</span>
where <span class="math inline">\(\phi_{ij}\)</span> denotes the kinship
coefficient between <span class="math inline">\(i\)</span>-th and <span class="math inline">\(j\)</span>-th sample. <span class="math inline">\(G_{im}\)</span> and <span class="math inline">\(G_{jm}\)</span> denotes genotypes of <span class="math inline">\(m\)</span>-th SNP for <span class="math inline">\(i\)</span>-th and <span class="math inline">\(j\)</span>-th sample. <span class="math inline">\(p_m\)</span> denotes allele frequency of <span class="math inline">\(m\)</span>-th SNP. <span class="math inline">\(R^2\)</span> is calculated as <span class="math inline">\(R^2 = \frac{Var(D)}{2q(1-q)}\)</span>, where <span class="math inline">\(D\)</span> is the dosage genotype. We classified
sample pairs as k-degree related if <span class="math inline">\(2^{-k-1.5} &lt; \phi_{ij} &lt; 2^{-k-0.5}\)</span>
(<a href="https://academic.oup.com/bioinformatics/article/26/22/2867/228512">Manichaikul
et al. Bioinformatics 2010</a>). A zero-degree related pair means
monozygotic twins (MZ) or duplicates. Sample pairs more distant than 3rd
degree are treated as unrelated.</p>
<p>Inbreeding coefficients are calculated as: <span class="math inline">\(F=2\phi_{ii}-1\)</span>, where <span class="math inline">\(\phi_{ii}\)</span> is the self-kinship coefficient
for sample <span class="math inline">\(i\)</span>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">getKinship</span>(dosage)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>kinship <span class="ot">&lt;-</span> res<span class="sc">$</span>kinship <span class="co"># kinship coefficients</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>inbreed <span class="ot">&lt;-</span> res<span class="sc">$</span>inbreed <span class="co"># inbreeding coefficients</span></span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
